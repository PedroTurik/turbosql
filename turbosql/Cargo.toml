[package]
authors = ["Eden <rust@eden2.com>"]
categories = ["database"]
description = "Easy local data persistence layer, backed by SQLite"
edition = "2018"
keywords = ["database", "sqlite", "sql"]
license = "MIT OR Apache-2.0"
name = "turbosql"
readme = "../README.md"
repository = "https://github.com/trevyn/turbosql"
version = "0.0.2"

[dependencies]
anyhow = "1"
directories-next = "2"
itertools = "0.10"
log = "0.4"
once_cell = "1.5"
rusqlite = {version = "0.24", features = ["bundled", "blob"]}
serde = {version = "1", features = ["derive"]}
toml = "0.5"
turbosql-impl = {path = "../turbosql-impl", version = "0.0.2"}

[dev-dependencies]
doc-comment = "0.3"
i54_ = {version = "0.0.1", features = ["rusqlite"]}

[features]
test = ["turbosql-impl/test"]

[[test]]
name = "integration_test"
path = "tests/integration_test.rs"
required-features = ["test"]
