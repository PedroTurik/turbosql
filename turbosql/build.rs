// this file must exist,
// so that the OUT_DIR env var is set on build,
// so that we can find the migrations.toml file,
// so that we can embed it in your binary,
// so that we can run your migrations.

fn main() {}
