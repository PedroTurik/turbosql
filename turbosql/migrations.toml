# This file represents your application's auto-generated SQLite schema.
# Normally it would be auto-populated from your struct definitions.
# But in the turbosql crate, it just exists so that include_str! doesn't error in various build scenarios.
