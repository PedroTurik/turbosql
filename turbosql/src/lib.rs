#[cfg(all(not(feature = "test"), any(test, doctest)))]
compile_error!("turbosql must be tested with '--features test'");

#[cfg(all(feature = "test", doctest))]
doc_comment::doctest!("../../README.md");

use itertools::{
 EitherOrBoth::{Both, Left, Right},
 Itertools,
};
use rusqlite::{Connection, OpenFlags};
use serde::Deserialize;
use std::cell::RefCell;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

// these re-exports are used in macro expansions

#[doc(hidden)]
pub use once_cell::sync::Lazy;
#[doc(hidden)]
pub use rusqlite::{
 params, types::FromSql, types::FromSqlResult, types::ToSql, types::ToSqlOutput, types::Value,
 types::ValueRef, Error, OptionalExtension, Result,
};
#[doc(hidden)]
pub use serde::Serialize;
pub use turbosql_impl::{execute, select, Turbosql};

/// Wrapper for `Vec<u8>` that may one day impl `Read`, `Write` and `Seek` traits.
pub type Blob = Vec<u8>;

#[derive(Clone, Debug, Deserialize, Default)]
struct MigrationsToml {
 migrations_append_only: Option<Vec<String>>,
 output_generated_schema_for_your_information_do_not_edit: Option<String>,
}

#[derive(Clone, Debug, Default)]
struct DbPath {
 path: PathBuf,
 opened: bool,
}

static __DB_PATH: Lazy<Mutex<DbPath>> = Lazy::new(|| {
 #[cfg(not(feature = "test"))]
 let path = {
  let exe_stem = std::env::current_exe().unwrap().file_stem().unwrap().to_owned();
  let exe_stem_lossy = exe_stem.to_string_lossy();

  let path = directories_next::ProjectDirs::from("org", &exe_stem_lossy, &exe_stem_lossy)
   .unwrap()
   .data_dir()
   .to_owned();

  std::fs::create_dir_all(&path).unwrap();

  path.join(exe_stem).with_extension("sqlite")
 };

 #[cfg(feature = "test")]
 let path = Path::new(":memory:").to_owned();

 Mutex::new(DbPath { path, ..Default::default() })
});

fn run_migrations(conn: &mut Connection) {
 cfg_if::cfg_if! {
  if #[cfg(doc)] {
   // if these are what's run in doctests, could add a test struct here to scaffold one-liner tests
   let toml_decoded: MigrationsToml = MigrationsToml::default();
  } else if #[cfg(feature = "test")] {
   let toml_decoded: MigrationsToml = toml::from_str(include_str!("../../test.migrations.toml")).unwrap();
  } else {
   let toml_decoded: MigrationsToml = toml::from_str(include_str!(concat!(env!("OUT_DIR"), "/../../../../../migrations.toml"))).expect("Unable to decode embedded migrations.toml");
  }
 };

 let target_migrations = toml_decoded.migrations_append_only.unwrap_or_else(Vec::new);

 // filter out comments
 let target_migrations: Vec<_> =
  target_migrations.into_iter().filter(|m| !m.starts_with("--")).collect();

 conn.execute("BEGIN EXCLUSIVE TRANSACTION", params![]).unwrap();

 let _ = conn.execute("ALTER TABLE turbosql_migrations RENAME TO _turbosql_migrations", params![]);

 let result = conn.query_row(
  "SELECT sql FROM sqlite_master WHERE name = ?",
  params!["_turbosql_migrations"],
  |row| {
   let sql: String = row.get(0).unwrap();
   Ok(sql)
  },
 );

 match result {
  Err(rusqlite::Error::QueryReturnedNoRows) => {
   // no migrations table exists yet, create
   conn
    .execute_batch(
     r#"CREATE TABLE _turbosql_migrations (rowid INTEGER PRIMARY KEY, migration TEXT NOT NULL)"#,
    )
    .expect("CREATE TABLE _turbosql_migrations");
  }
  Err(err) => {
   panic!("Could not query sqlite_master table: {}", err);
  }
  Ok(_) => (),
 }

 let applied_migrations = conn
  .prepare("SELECT migration FROM _turbosql_migrations ORDER BY rowid")
  .unwrap()
  .query_map(params![], |row| Ok(row.get(0).unwrap()))
  .unwrap()
  .map(|x: Result<String, _>| x.unwrap())
  .filter(|m| !m.starts_with("--"))
  .collect::<Vec<String>>();

 // execute migrations

 applied_migrations.iter().zip_longest(&target_migrations).for_each(|item| match item {
  Both(a, b) => {
   if a != b {
    panic!("Mismatch in Turbosql migrations! {:?} != {:?}", a, b)
   }
  }
  Left(_) => panic!("More migrations are applied than target"),
  Right(migration) => {
   // eprintln!("insert -> {:#?}", migration);
   if !migration.starts_with("--") {
    conn.execute(migration, params![]).unwrap();
   }
   conn
    .execute("INSERT INTO _turbosql_migrations(migration) VALUES(?)", params![migration])
    .unwrap();
  }
 });

 // TODO: verify schema against output_generated_schema_for_your_information_do_not_edit

 //    if sql != create_sql {
 //     println!("{}", sql);
 //     println!("{}", create_sql);
 //     panic!("Turbosql sqlite schema does not match! Delete database file to continue.");
 //    }

 conn.execute("COMMIT", params![]).unwrap();
}

fn open_db() -> Connection {
 let start = std::time::Instant::now();

 let mut db_path = __DB_PATH.lock().unwrap();

 // We are handling the mutex by being thread_local, so SQLite can be opened in no-mutex mode; see:
 // http://sqlite.1065341.n5.nabble.com/SQLITE-OPEN-FULLMUTEX-vs-SQLITE-OPEN-NOMUTEX-td104785.html

 let mut conn = Connection::open_with_flags(
  &db_path.path,
  OpenFlags::SQLITE_OPEN_READ_WRITE
   | OpenFlags::SQLITE_OPEN_CREATE
   | OpenFlags::SQLITE_OPEN_NO_MUTEX,
 )
 .expect("rusqlite::Connection::open_with_flags");

 conn
  .execute_batch(
   r#"
    PRAGMA busy_timeout=3000;
    PRAGMA auto_vacuum=INCREMENTAL;
    PRAGMA journal_mode=WAL;
    PRAGMA wal_autocheckpoint=8000;
    PRAGMA synchronous=NORMAL;
   "#,
  )
  .expect("Execute PRAGMAs");

 if !db_path.opened {
  run_migrations(&mut conn);
  db_path.opened = true;
 }

 log::info!("db opened in {:?}", start.elapsed());

 conn
}

#[doc(hidden)]
thread_local! {
 pub static __TURBOSQL_DB: RefCell<Connection> = RefCell::new(open_db());
}

/// Set the local path and filename where Turbosql will store the underlying SQLite database.
///
/// Must be called before any usage of Turbosql macros or will return an error.
pub fn set_db_path(path: &Path) -> Result<(), anyhow::Error> {
 let mut db_path = __DB_PATH.lock().unwrap();

 if db_path.opened {
  return Err(anyhow::anyhow!("Trying to set path when DB is already opened"));
 }

 db_path.path = path.to_owned();

 Ok(())
}
