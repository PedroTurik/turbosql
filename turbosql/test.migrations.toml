# This file is auto-generated by Turbosql.
# It is used to create and apply automatic schema migrations.
# It should be checked into source control.
# Modifying it by hand may be dangerous; see the docs.

migrations_append_only = [
  'CREATE TABLE person (rowid INTEGER PRIMARY KEY) STRICT',
  'ALTER TABLE person ADD COLUMN name TEXT',
  'ALTER TABLE person ADD COLUMN age INTEGER',
  'ALTER TABLE person ADD COLUMN image_jpg BLOB',
]
output_generated_schema_for_your_information_do_not_edit = '''
  CREATE TABLE _turbosql_migrations (
    rowid INTEGER PRIMARY KEY,
    migration TEXT NOT NULL
  ) STRICT
  CREATE TABLE person (
    rowid INTEGER PRIMARY KEY,
    name TEXT,
    age INTEGER,
    image_jpg BLOB
  ) STRICT
'''
[output_generated_tables_do_not_edit.person]
name = 'person'

[[output_generated_tables_do_not_edit.person.columns]]
name = 'rowid'
rust_type = 'Option < i64 >'
sql_type = 'INTEGER PRIMARY KEY'

[[output_generated_tables_do_not_edit.person.columns]]
name = 'name'
rust_type = 'Option < String >'
sql_type = 'TEXT'

[[output_generated_tables_do_not_edit.person.columns]]
name = 'age'
rust_type = 'Option < i64 >'
sql_type = 'INTEGER'

[[output_generated_tables_do_not_edit.person.columns]]
name = 'image_jpg'
rust_type = 'Option < Blob >'
sql_type = 'BLOB'
