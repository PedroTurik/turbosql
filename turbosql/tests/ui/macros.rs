// TRYBUILD=overwrite cargo test -p turbosql --test ui

use turbosql::{execute, select, Turbosql};

#[derive(Turbosql, Default)]
struct NoOption {
	rowid: Option<i64>,
	e: u8,
}

#[derive(Turbosql, Default)]
struct NoRowId {
	age: Option<u8>,
}

#[derive(Turbosql, Default)]
struct U64 {
	rowid: Option<i64>,
	e: Option<u64>,
}

#[derive(Turbosql, Default)]
struct Person {
	rowid: Option<i64>,
	name: Option<String>,
	age: Option<u8>,
}

fn main() {
	select!(Person "WHERE age = " 24 " AND name = ?", "Bob").unwrap();
	select!(Person "WHERE age = " 24 " AND name = $name").unwrap();
	select!(Person "WHERE age = ?", 1, 2).unwrap();
	select!(Person "WHERE age = ").unwrap();
	select!("UPDATE person SET age = 1").unwrap();
	execute!("SELECT 1").unwrap();
	select!(Person "WHERE nonexistentcolumn = 1").unwrap();
	select!(Nonexistenttable).unwrap();
	select!(Vec).unwrap();
	select!(Vec<"what">).unwrap();
}
