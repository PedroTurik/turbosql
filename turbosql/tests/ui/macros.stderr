error: Turbosql types must be wrapped in Option for forward/backward schema compatibility. Try: Option<u8>
 --> tests/ui/macros.rs:6:5
  |
6 |  e: u8,
  |     ^^

error: derive(Turbosql) structs must include a 'rowid: Option<i64>' field
 --> tests/ui/macros.rs:9:10
  |
9 | #[derive(Turbosql, Default)]
  |          ^^^^^^^^
  |
  = note: this error originates in the derive macro `Turbosql` (in Nightly builds, run with -Z macro-backtrace for more info)

error: SQLite cannot natively store unsigned 64-bit integers, so Turbosql does not support u64 fields. Use i64, u32, f64, or a string or binary format instead. (see https://github.com/trevyn/turbosql/issues/3 )
  --> tests/ui/macros.rs:17:5
   |
17 |  e: Option<u64>,
   |     ^^^^^^^^^^^

error: expected expression
  --> tests/ui/macros.rs:28:50
   |
28 |  select!(Person "WHERE age = " 24 " AND name = ?", "Bob").unwrap();
   |                                                  ^

error: SQLite named parameters not currently supported.
  --> tests/ui/macros.rs:29:2
   |
29 |  select!(Person "WHERE age = " 24 " AND name = $name").unwrap();
   |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `select` (in Nightly builds, run with -Z macro-backtrace for more info)

error: Expected 1 bound parameter, got 2: "SELECT rowid, name, age FROM person WHERE age = ?"
  --> tests/ui/macros.rs:30:17
   |
30 |  select!(Person "WHERE age = ?", 1, 2).unwrap();
   |                 ^^^^^^^^^^^^^^^^^^^^^

error: Error validating SQL statement: "incomplete input". SQL: "SELECT rowid, name, age FROM person WHERE age = "
  --> tests/ui/macros.rs:31:2
   |
31 |  select!(Person "WHERE age = ").unwrap();
   |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `select` (in Nightly builds, run with -Z macro-backtrace for more info)

error: No rows returned from SQL, use execute! instead.
  --> tests/ui/macros.rs:32:2
   |
32 |  select!("UPDATE person SET age = 1").unwrap();
   |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `select` (in Nightly builds, run with -Z macro-backtrace for more info)

error: Rows returned from SQL, use select! instead.
  --> tests/ui/macros.rs:33:2
   |
33 |  execute!("SELECT 1").unwrap();
   |  ^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `execute` (in Nightly builds, run with -Z macro-backtrace for more info)

error: Error validating SQL statement: "no such column: nonexistentcolumn in SELECT rowid, name, age FROM person WHERE nonexistentcolumn = 1 at offset 42". SQL: "SELECT rowid, name, age FROM person WHERE nonexistentcolumn = 1"
  --> tests/ui/macros.rs:34:2
   |
34 |  select!(Person "WHERE nonexistentcolumn = 1").unwrap();
   |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `select` (in Nightly builds, run with -Z macro-backtrace for more info)

error: Table "nonexistenttable" not found. Does struct Nonexistenttable exist and have #[derive(Turbosql, Default)]?
  --> tests/ui/macros.rs:35:10
   |
35 |  select!(Nonexistenttable).unwrap();
   |          ^^^^^^^^^^^^^^^^

error: expected `<`
  --> tests/ui/macros.rs:36:2
   |
36 |  select!(Vec).unwrap();
   |  ^^^^^^^^^^^^
   |
   = note: this error originates in the macro `select` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected identifier
  --> tests/ui/macros.rs:37:14
   |
37 |  select!(Vec<"what">).unwrap();
   |              ^^^^^^
